src/rgw/rgw_lua_filter.cc

```diff
@@ -0,0 +1,138 @@
+    } else {
+      lua_pushnil(L);
+    }
+    
+    return ONE_RETURNVAL;
+  }
+
+  static int PairsClosure(lua_State* L) {
+    auto bl = reinterpret_cast<bufferlist*>(lua_touserdata(L, lua_upvalueindex(1)));
+    auto dpp = reinterpret_cast<DoutPrefixProvider*>(lua_touserdata(L, lua_upvalueindex(2)));
+    ceph_assert(bl);
+    lua_pushlightuserdata(L, bl);
+    lua_pushlightuserdata(L, dpp);
+    lua_pushcclosure(L, stateless_iter, TWO_UPVALS); // push the stateless iterator function
+    lua_pushnil(L);                                 // indicate this is the first call
+    // return stateless_iter, nil
+    ldpp_dout(dpp, 20) << "Lua: BufferlistMetaTable::PairsClosure called" << dendl;
+
+    return TWO_RETURNVALS;
+  }
+  
+  static int stateless_iter(lua_State* L) {
+    // based on: http://lua-users.org/wiki/GeneralizedPairsAndIpairs
+    auto bl = reinterpret_cast<bufferlist*>(lua_touserdata(L, lua_upvalueindex(1)));
+    auto dpp = reinterpret_cast<DoutPrefixProvider*>(lua_touserdata(L, lua_upvalueindex(2)));
+    lua_Integer index;
+    if (lua_isnil(L, -1)) {
+      index = 1;
+      ldpp_dout(dpp, 20) << "Lua: BufferlistMetaTable::stateless_iter() first call index =" << index << dendl;
+    } else {
+      index = luaL_checkinteger(L, -1) + 1;
+    }
+    ldpp_dout(dpp, 20) << "Lua: BufferlistMetaTable::stateless_iter() index = " << index << dendl;
+
+    // lua arrays start from 1
+    auto it = bl->begin(index-1);
+
+    if (index > bl->length()) {
+      // index of the last element was provided
+      ldpp_dout(dpp, 20) << "Lua: BufferlistMetaTable::stateless_iter() index of last element was provided" << dendl;
+      lua_pushnil(L);
+      lua_pushnil(L);
+      // return nil, nil
+    } else {
+      ldpp_dout(dpp, 20) << "Lua: BufferlistMetaTable::stateless_iter() returning index " << index << " and value" << dendl;
+      lua_pushinteger(L, index);
+      push_bufferlist_byte(L, it);
+      // return key, value
+    }
+
+    return TWO_RETURNVALS;
+  }
+  
+  static int LenClosure(lua_State* L) {
+    const auto bl = reinterpret_cast<bufferlist*>(lua_touserdata(L, lua_upvalueindex(1)));
+    auto dpp = reinterpret_cast<DoutPrefixProvider*>(lua_touserdata(L, lua_upvalueindex(2)));
+
+    ldpp_dout(dpp, 20) << "Lua: BufferlistMetaTable::LenClosure() returning length " << bl->length() << " and value" << dendl;
+    lua_pushinteger(L, bl->length());
+
+    return ONE_RETURNVAL;
+  }
+};
+
+int RGWObjFilter::execute(bufferlist& bl) const {
+  auto L = luaL_newstate();
+  lua_state_guard lguard(L);
+
+  open_standard_libs(L);
+
+  create_debug_action(L, cct);  
+
+  create_metatable<BufferlistMetaTable>(L, true, &bl, const_cast<DoutPrefixProvider*>(dpp));
+
+  try {
+    // execute the lua script
+    if (luaL_dostring(L, script.c_str()) != LUA_OK) {
+      const std::string err(lua_tostring(L, -1));
+      ldpp_dout(dpp, 1) << "Lua ERROR: " << err << dendl;
+    }
+  } catch (const std::runtime_error& e) {
```

Comment:
what throws `runtime_error` here?

if you want the ability to write lua scripts that can block/reject a PUT request based on its payload, how should the script communicate that? i'm guessing that we'd just return that error back to `RGWGet/PutObjFilter`

Reply:
> what throws `runtime_error` here?

any C++ code invoked from inside the lua script

> if you want the ability to write lua scripts that can block/reject a PUT request based on its payload, how should the script communicate that? i'm guessing that we'd just return that error back to `RGWGet/PutObjFilter`

currently, I ignore any Lua errors (so that a lua script with an error will not block all requests).
however, this is probably a good idea to add a parameter that could be set from lua to block the request

Reply:
well. there is a long discussion on how to properly handle exceptions thrown in c++ code invoked by lua:
http://lua-users.org/wiki/ErrorHandlingBetweenLuaAndCplusplus

one simple takeaway is that we don't need any `try/catch` here (and in the other places we invoke lua), since the lua code will not pass any exception thrown from c++.
but there are other issues that we probably need to handle:
* what will happen with these exceptions?
* if the c++ code invoked by lua uses any resources, and a lua error happens, dtors will be skipped. we need to make sure that we don't have such resource leaks

(this will be done in a separate PR, as this is unrelated to the object data access)